File: common/src/main/java/dev/thomasqtruong/veryscuffedcobblemonbreeding/util/PokemonUtility.java
package dev.thomasqtruong.veryscuffedcobblemonbreeding.util;

import com.cobblemon.mod.common.api.pokemon.stats.Stats;
import com.cobblemon.mod.common.api.properties.CustomPokemonProperty;
import com.cobblemon.mod.common.api.properties.CustomPokemonPropertyType;
import com.cobblemon.mod.common.item.PokemonItem;
import com.cobblemon.mod.common.pokemon.Pokemon;
import com.cobblemon.mod.common.util.LocalizationUtilsKt;
import net.minecraft.item.ItemStack;
import net.minecraft.text.*;
import net.minecraft.util.Formatting;

import java.util.List;

public class PokemonUtility {


    public static Text getHoverText(MutableText toSend, Pokemon pokemon) {
        MutableText statsHoverText = Text.literal("").fillStyle(Style.EMPTY);
        statsHoverText.append(pokemon.getDisplayName().setStyle(Style.EMPTY.withColor(Formatting.DARK_GREEN).withUnderline(true)));
        if (pokemon.getShiny()) {
            statsHoverText.append(Text.literal(" ★").formatted(Formatting.GOLD));
        }
        statsHoverText.append(Text.literal("\n"));
        statsHoverText.append(Text.literal("Gender: ").formatted(Formatting.GREEN).append(Text.literal(String.valueOf(pokemon.getGender())).formatted(Formatting.WHITE)));
        statsHoverText.append(Text.literal("\n"));
        statsHoverText.append(Text.literal("Level: ").formatted(Formatting.AQUA).append(Text.literal(String.valueOf(pokemon.getLevel())).formatted(Formatting.WHITE)));
        statsHoverText.append(Text.literal("\n"));
        statsHoverText.append(Text.literal("Nature: ").formatted(Formatting.YELLOW).append(LocalizationUtilsKt.lang(pokemon.getNature().getDisplayName().replace("cobblemon.", "")).formatted(Formatting.WHITE)));
        statsHoverText.append(Text.literal("\n"));
        statsHoverText.append(Text.literal("Ability: ").formatted(Formatting.GOLD).append(LocalizationUtilsKt.lang(pokemon.getAbility().getDisplayName().replace("cobblemon.", "")).formatted(Formatting.WHITE)));
        statsHoverText.append(Text.literal("\n"));
        statsHoverText.append(Text.literal("Form: ").formatted(Formatting.GREEN).append(Text.literal(pokemon.getForm().getName()).formatted(Formatting.WHITE)));

        Text statsText = Texts.join(Text.literal("[Stats]").formatted(Formatting.RED)
                .getWithStyle(Style.EMPTY.withColor(Formatting.RED)
                        .withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, statsHoverText))), Text.of(""));

        toSend.append(statsText);

        MutableText evsText = Text.literal(" [EVs]").formatted(Formatting.GOLD);
        MutableText evsHoverText = Text.literal("");
        List<Text> evsHoverTextList = Text.literal("EVs").formatted(Formatting.GOLD).getWithStyle(Style.EMPTY.withUnderline(Boolean.TRUE));
        evsHoverTextList.add(Text.literal("\n"));
        evsHoverTextList.add(Text.literal("HP: ").formatted(Formatting.RED).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.HP))).formatted(Formatting.WHITE)));
        evsHoverTextList.add(Text.literal("\n"));
        evsHoverTextList.add(Text.literal("Attack: ").formatted(Formatting.BLUE).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.ATTACK))).formatted(Formatting.WHITE)));
        evsHoverTextList.add(Text.literal("\n"));
        evsHoverTextList.add(Text.literal("Defense: ").formatted(Formatting.GRAY).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.DEFENCE))).formatted(Formatting.WHITE)));
        evsHoverTextList.add(Text.literal("\n"));
        evsHoverTextList.add(Text.literal("Sp. Attack: ").formatted(Formatting.AQUA).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.SPECIAL_ATTACK))).formatted(Formatting.WHITE)));
        evsHoverTextList.add(Text.literal("\n"));
        evsHoverTextList.add(Text.literal("Sp. Defense: ").formatted(Formatting.YELLOW).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.SPECIAL_DEFENCE))).formatted(Formatting.WHITE)));
        evsHoverTextList.add(Text.literal("\n"));
        evsHoverTextList.add(Text.literal("Speed: ").formatted(Formatting.GREEN).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.SPEED))).formatted(Formatting.WHITE)));

        evsHoverTextList.forEach(evsHoverText::append);
        List<Text> evsList = evsText.getWithStyle(evsText.getStyle().withHoverEvent(
                new HoverEvent(HoverEvent.Action.SHOW_TEXT,
                        evsHoverText
                ))
        );
        evsList.forEach(toSend::append);

        MutableText ivsText = Text.literal(" [IVs]").formatted(Formatting.LIGHT_PURPLE);
        MutableText ivsHoverText = Text.literal("");
        List<Text> ivsHoverTextList = Text.literal("IVs").formatted(Formatting.GOLD).getWithStyle(Style.EMPTY.withUnderline(Boolean.TRUE));
        ivsHoverTextList.add(Text.literal("\n"));
        ivsHoverTextList.add(Text.literal("HP: ").formatted(Formatting.RED).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.HP))).formatted(Formatting.WHITE)));
        ivsHoverTextList.add(Text.literal("\n"));
        ivsHoverTextList.add(Text.literal("Attack: ").formatted(Formatting.BLUE).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.ATTACK))).formatted(Formatting.WHITE)));
        ivsHoverTextList.add(Text.literal("\n"));
        ivsHoverTextList.add(Text.literal("Defense: ").formatted(Formatting.GRAY).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.DEFENCE))).formatted(Formatting.WHITE)));
        ivsHoverTextList.add(Text.literal("\n"));
        ivsHoverTextList.add(Text.literal("Sp. Attack: ").formatted(Formatting.AQUA).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.SPECIAL_ATTACK))).formatted(Formatting.WHITE)));
        ivsHoverTextList.add(Text.literal("\n"));
        ivsHoverTextList.add(Text.literal("Sp. Defense: ").formatted(Formatting.YELLOW).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.SPECIAL_DEFENCE))).formatted(Formatting.WHITE)));
        ivsHoverTextList.add(Text.literal("\n"));
        ivsHoverTextList.add(Text.literal("Speed: ").formatted(Formatting.GREEN).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.SPEED))).formatted(Formatting.WHITE)));

        ivsHoverTextList.forEach(ivsHoverText::append);
        List<Text> ivsList = ivsText.getWithStyle(ivsText.getStyle().withHoverEvent(
                new HoverEvent(HoverEvent.Action.SHOW_TEXT,
                        ivsHoverText
                ))
        );
        ivsList.forEach(toSend::append);

        MutableText movesText = Text.literal(" [Moves]").formatted(Formatting.BLUE);
        MutableText movesHoverText = Text.literal("");
        List<Text> movesHoverTextList = Text.literal("Moves").formatted(Formatting.BLUE).getWithStyle(Style.EMPTY.withUnderline(Boolean.TRUE));
        movesHoverTextList.add(Text.literal("\n"));
        String moveOne = pokemon.getMoveSet().getMoves().size() >= 1 ? pokemon.getMoveSet().get(0).getDisplayName().getString() : "Empty";
        String moveTwo = pokemon.getMoveSet().getMoves().size() >= 2 ? pokemon.getMoveSet().get(1).getDisplayName().getString() : "Empty";
        String moveThree = pokemon.getMoveSet().getMoves().size() >= 3 ? pokemon.getMoveSet().get(2).getDisplayName().getString() : "Empty";
        String moveFour = pokemon.getMoveSet().getMoves().size() >= 4 ? pokemon.getMoveSet().get(3).getDisplayName().getString() : "Empty";
        movesHoverTextList.add(Text.literal("Move 1: ").formatted(Formatting.RED).append(Text.literal(moveOne).formatted(Formatting.WHITE)));
        movesHoverTextList.add(Text.literal("\n"));
        movesHoverTextList.add(Text.literal("Move 2: ").formatted(Formatting.YELLOW).append(Text.literal(moveTwo).formatted(Formatting.WHITE)));
        movesHoverTextList.add(Text.literal("\n"));
        movesHoverTextList.add(Text.literal("Move 3: ").formatted(Formatting.AQUA).append(Text.literal(moveThree).formatted(Formatting.WHITE)));
        movesHoverTextList.add(Text.literal("\n"));
        movesHoverTextList.add(Text.literal("Move 4: ").formatted(Formatting.GREEN).append(Text.literal(moveFour).formatted(Formatting.WHITE)));

        movesHoverTextList.forEach(movesHoverText::append);
        List<Text> movesList = movesText.getWithStyle(movesText.getStyle().withHoverEvent(
                new HoverEvent(HoverEvent.Action.SHOW_TEXT,
                        movesHoverText
                ))
        );
        movesList.forEach(toSend::append);
        return toSend;
    }

    public static ItemStack pokemonToItem(Pokemon pokemon) {
        String moveOne = pokemon.getMoveSet().getMoves().size() >= 1 ? pokemon.getMoveSet().get(0).getDisplayName().getString() : "Empty";
        String moveTwo = pokemon.getMoveSet().getMoves().size() >= 2 ? pokemon.getMoveSet().get(1).getDisplayName().getString() : "Empty";
        String moveThree = pokemon.getMoveSet().getMoves().size() >= 3 ? pokemon.getMoveSet().get(2).getDisplayName().getString() : "Empty";
        String moveFour = pokemon.getMoveSet().getMoves().size() >= 4 ? pokemon.getMoveSet().get(3).getDisplayName().getString() : "Empty";

        ItemStack itemstack = new ItemBuilder(PokemonItem.from(pokemon, 1))
                .hideAdditional()
                .addLore(new Text[]{
                        pokemon.getCaughtBall().item().getName().copy().setStyle(Style.EMPTY.withItalic(true).withColor(Formatting.DARK_GRAY)),
                        Text.literal("Gender: ").formatted(Formatting.GREEN).append(Text.literal(String.valueOf(pokemon.getGender())).formatted(Formatting.WHITE)),
                        Text.literal("Level: ").formatted(Formatting.AQUA).append(Text.literal(String.valueOf(pokemon.getLevel())).formatted(Formatting.WHITE)),
                        Text.literal("Nature: ").formatted(Formatting.YELLOW).append(LocalizationUtilsKt.lang(pokemon.getNature().getDisplayName().replace("cobblemon.", "")).formatted(Formatting.WHITE)),
                        Text.literal("Ability: ").formatted(Formatting.GOLD).append(LocalizationUtilsKt.lang(pokemon.getAbility().getDisplayName().replace("cobblemon.", "")).formatted(Formatting.WHITE)),
                        Text.literal("IVs: ").formatted(Formatting.LIGHT_PURPLE),
                        Text.literal("  HP: ").formatted(Formatting.RED).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.HP))).formatted(Formatting.WHITE))
                                .append(Text.literal("  Atk: ").formatted(Formatting.BLUE).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.ATTACK))).formatted(Formatting.WHITE)))
                                .append(Text.literal("  Def: ").formatted(Formatting.GRAY).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.DEFENCE))).formatted(Formatting.WHITE))),
                        Text.literal("  SpAtk: ").formatted(Formatting.AQUA).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.SPECIAL_ATTACK))).formatted(Formatting.WHITE))
                                .append(Text.literal("  SpDef: ").formatted(Formatting.YELLOW).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.SPECIAL_DEFENCE))).formatted(Formatting.WHITE)))
                                .append(Text.literal("  Spd: ").formatted(Formatting.GREEN).append(Text.literal(String.valueOf(pokemon.getIvs().getOrDefault(Stats.SPEED))).formatted(Formatting.WHITE))),

                        Text.literal("EVs: ").formatted(Formatting.DARK_AQUA),
                        Text.literal("  HP: ").formatted(Formatting.RED).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.HP))).formatted(Formatting.WHITE))
                                .append(Text.literal("  Atk: ").formatted(Formatting.BLUE).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.ATTACK))).formatted(Formatting.WHITE)))
                                .append(Text.literal("  Def: ").formatted(Formatting.GRAY).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.DEFENCE))).formatted(Formatting.WHITE))),
                        Text.literal("  SpAtk: ").formatted(Formatting.AQUA).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.SPECIAL_ATTACK))).formatted(Formatting.WHITE))
                                .append(Text.literal("  SpDef: ").formatted(Formatting.YELLOW).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.SPECIAL_DEFENCE))).formatted(Formatting.WHITE)))
                                .append(Text.literal("  Spd: ").formatted(Formatting.GREEN).append(Text.literal(String.valueOf(pokemon.getEvs().getOrDefault(Stats.SPEED))).formatted(Formatting.WHITE))),
                        Text.literal("Moves: ").formatted(Formatting.DARK_GREEN),
                        Text.literal(" ").append(Text.literal(moveOne).formatted(Formatting.WHITE)),
                        Text.literal(" ").append(Text.literal(moveTwo).formatted(Formatting.WHITE)),
                        Text.literal(" ").append(Text.literal(moveThree).formatted(Formatting.WHITE)),
                        Text.literal(" ").append(Text.literal(moveFour).formatted(Formatting.WHITE)),
                        Text.literal("Form: ").formatted(Formatting.GOLD).append(pokemon.getForm().getName())
                })
                .setCustomName(
                        pokemon.getShiny() ?
                                pokemon.getDisplayName().formatted(Formatting.GRAY).append(Text.literal(" ★").formatted(Formatting.GOLD)) :
                                pokemon.getDisplayName().formatted(Formatting.GRAY)
                )
                .build();
        return itemstack;
    }
}

File: common/src/main/kotlin/dev/thomasqtruong/veryscuffedcobblemonbreeding/util/PokemonUtility.kt
package dev.thomasqtruong.veryscuffedcobblemonbreeding.util

import com.cobblemon.mod.common.api.pokemon.stats.Stats
import com.cobblemon.mod.common.item.PokemonItem
import com.cobblemon.mod.common.pokemon.Pokemon
import com.cobblemon.mod.common.util.lang
import net.minecraft.ChatFormatting
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.HoverEvent
import net.minecraft.network.chat.MutableComponent
import net.minecraft.network.chat.Style
import net.minecraft.world.item.ItemStack
import java.lang.Boolean
import kotlin.arrayOf

object PokemonUtility {
    fun getHoverText(toSend: MutableComponent, pokemon: Pokemon): Component {
        val statsHoverText = Component.literal("").withStyle(Style.EMPTY)

        if (pokemon.shiny) {
            statsHoverText.append(Component.literal(" ★").withStyle(ChatFormatting.GOLD))
        }
        statsHoverText.append(pokemon.species.translatedName.copy().withStyle(Style.EMPTY.withColor(ChatFormatting.DARK_GREEN).withUnderlined(true)))
        statsHoverText.append(Component.literal("\n"))
        if (pokemon.nickname != null) {
            statsHoverText.append(Component.literal("Nickname: ").withStyle(ChatFormatting.GRAY).append(Component.literal(pokemon.nickname!!.string).withStyle(ChatFormatting.WHITE)))
            statsHoverText.append(Component.literal("\n"))
        }
        statsHoverText.append(Component.literal("Level: ").withStyle(ChatFormatting.AQUA).append(Component.literal(pokemon.level.toString()).withStyle(ChatFormatting.WHITE)))
        statsHoverText.append(Component.literal("\n"))
        statsHoverText.append(Component.literal("Nature: ").withStyle(ChatFormatting.YELLOW).append(lang(pokemon.nature.displayName.replace("cobblemon.", "")).withStyle(ChatFormatting.WHITE)))
        statsHoverText.append(Component.literal("\n"))
        statsHoverText.append(Component.literal("Ability: ").withStyle(ChatFormatting.GOLD).append(lang(pokemon.ability.displayName.replace("cobblemon.", "")).withStyle(ChatFormatting.WHITE)))
        statsHoverText.append(Component.literal("\n"))
        statsHoverText.append(Component.literal("Form: ").withStyle(ChatFormatting.GREEN).append(Component.literal(pokemon.form.name).withStyle(ChatFormatting.WHITE)))

        val statsText = Component.literal("[Stats]").withStyle(Style.EMPTY.withColor(ChatFormatting.RED)
            .withHoverEvent(HoverEvent(HoverEvent.Action.SHOW_TEXT, statsHoverText)))

        toSend.append(statsText)

        // Add up Evs, round to 2 decimal places, multiply by 100 to get percent
        val allEvs = (pokemon.evs.getOrDefault(Stats.HP) + pokemon.evs.getOrDefault(Stats.ATTACK) + pokemon.evs.getOrDefault(Stats.DEFENCE) + pokemon.evs.getOrDefault(Stats.SPECIAL_ATTACK) + pokemon.evs.getOrDefault(Stats.SPECIAL_DEFENCE) + pokemon.evs.getOrDefault(Stats.SPEED)).toDouble()
        val avg = (allEvs / (510.0))
        val evPercent = Math.round(avg * 10000).toDouble() / 100

        val evsText = Component.literal("[EVs]").withStyle(ChatFormatting.GOLD)
        val evsHoverText = Component.literal("").withStyle(Style.EMPTY)
        evsHoverText.append(Component.literal("EVs ($evPercent%) ").withStyle(Style.EMPTY.withColor(ChatFormatting.GOLD).withUnderlined(Boolean.TRUE)))
        evsHoverText.append(Component.literal("\n").withStyle(Style.EMPTY.withUnderlined(false)))
        evsHoverText.append(Component.literal("HP: ").withStyle(ChatFormatting.RED).append(Component.literal(pokemon.evs.getOrDefault(Stats.HP).toString()).withStyle(ChatFormatting.WHITE)))
        evsHoverText.append(Component.literal("\n"))
        evsHoverText.append(Component.literal("Attack: ").withStyle(ChatFormatting.BLUE).append(Component.literal(pokemon.evs.getOrDefault(Stats.ATTACK).toString()).withStyle(ChatFormatting.WHITE)))
        evsHoverText.append(Component.literal("\n"))
        evsHoverText.append(Component.literal("Defense: ").withStyle(ChatFormatting.GRAY).append(Component.literal(pokemon.evs.getOrDefault(Stats.DEFENCE).toString()).withStyle(ChatFormatting.WHITE)))
        evsHoverText.append(Component.literal("\n"))
        evsHoverText.append(Component.literal("Sp. Attack: ").withStyle(ChatFormatting.AQUA).append(Component.literal(pokemon.evs.getOrDefault(Stats.SPECIAL_ATTACK).toString()).withStyle(ChatFormatting.WHITE)))
        evsHoverText.append(Component.literal("\n"))
        evsHoverText.append(Component.literal("Sp. Defense: ").withStyle(ChatFormatting.YELLOW).append(Component.literal(pokemon.evs.getOrDefault(Stats.SPECIAL_DEFENCE).toString()).withStyle(ChatFormatting.WHITE)))
        evsHoverText.append(Component.literal("\n"))
        evsHoverText.append(Component.literal("Speed: ").withStyle(ChatFormatting.GREEN).append(Component.literal(pokemon.evs.getOrDefault(Stats.SPEED).toString()).withStyle(ChatFormatting.WHITE)))
        evsText.setStyle(evsText.style.withHoverEvent(
            HoverEvent(HoverEvent.Action.SHOW_TEXT,
                evsHoverText
            ))
        )
        toSend.append(evsText)

        // Add up Ivs, take average, round to 2 decimal places, multiply by 100 to get percent
        val allIvs = pokemon.ivs.getOrDefault(Stats.HP) + pokemon.ivs.getOrDefault(Stats.ATTACK) + pokemon.ivs.getOrDefault(Stats.DEFENCE) + pokemon.ivs.getOrDefault(Stats.SPECIAL_ATTACK) + pokemon.ivs.getOrDefault(Stats.SPECIAL_DEFENCE) + pokemon.ivs.getOrDefault(Stats.SPEED)
        val ivAvg = (allIvs / (186.0))
        val ivPercent = Math.round(ivAvg * 10000).toDouble() / 100

        val ivsText = Component.literal("[IVs]").withStyle(ChatFormatting.LIGHT_PURPLE)
        val ivsHoverText = Component.literal("").withStyle(Style.EMPTY)
        ivsHoverText.append(Component.literal("IVs ($ivPercent%)").withStyle(Style.EMPTY.withColor(ChatFormatting.LIGHT_PURPLE).withUnderlined(true)))
        ivsHoverText.append(Component.literal("\n").withStyle(Style.EMPTY.withUnderlined(false)))
        ivsHoverText.append(Component.literal("HP: ").withStyle(ChatFormatting.RED).append(Component.literal(pokemon.ivs.getOrDefault(Stats.HP).toString()).withStyle(ChatFormatting.WHITE)))
        ivsHoverText.append(Component.literal("\n"))
        ivsHoverText.append(Component.literal("Attack: ").withStyle(ChatFormatting.BLUE).append(Component.literal(pokemon.ivs.getOrDefault(Stats.ATTACK).toString()).withStyle(ChatFormatting.WHITE)))
        ivsHoverText.append(Component.literal("\n"))
        ivsHoverText.append(Component.literal("Defense: ").withStyle(ChatFormatting.GRAY).append(Component.literal(pokemon.ivs.getOrDefault(Stats.DEFENCE).toString()).withStyle(ChatFormatting.WHITE)))
        ivsHoverText.append(Component.literal("\n"))
        ivsHoverText.append(Component.literal("Sp. Attack: ").withStyle(ChatFormatting.AQUA).append(Component.literal(pokemon.ivs.getOrDefault(Stats.SPECIAL_ATTACK).toString()).withStyle(ChatFormatting.WHITE)))
        ivsHoverText.append(Component.literal("\n"))
        ivsHoverText.append(Component.literal("Sp. Defense: ").withStyle(ChatFormatting.YELLOW).append(Component.literal(pokemon.ivs.getOrDefault(Stats.SPECIAL_DEFENCE).toString()).withStyle(ChatFormatting.WHITE)))
        ivsHoverText.append(Component.literal("\n"))
        ivsHoverText.append(Component.literal("Speed: ").withStyle(ChatFormatting.GREEN).append(Component.literal(pokemon.ivs.getOrDefault(Stats.SPEED).toString()).withStyle(ChatFormatting.WHITE)))
        ivsText.setStyle(ivsText.style.withHoverEvent(
            HoverEvent(HoverEvent.Action.SHOW_TEXT,
                ivsHoverText
            ))
        )
        toSend.append(ivsText)

        val movesText = Component.literal("[Moves]").withStyle(ChatFormatting.BLUE)
        val movesHoverText = Component.literal("").withStyle(Style.EMPTY)
        movesHoverText.append(Component.literal("Moves").withStyle(Style.EMPTY.withColor(ChatFormatting.BLUE).withUnderlined(Boolean.TRUE)))
        movesHoverText.append(Component.literal("\n").withStyle(Style.EMPTY.withUnderlined(false)))
        val moveOne = if (pokemon.moveSet.getMoves().size >= 1) pokemon.moveSet[0]!!.displayName.string else "Empty"
        val moveTwo = if (pokemon.moveSet.getMoves().size >= 2) pokemon.moveSet[1]!!.displayName.string else "Empty"
        val moveThree = if (pokemon.moveSet.getMoves().size >= 3) pokemon.moveSet[2]!!.displayName.string else "Empty"
        val moveFour = if (pokemon.moveSet.getMoves().size >= 4) pokemon.moveSet[3]!!.displayName.string else "Empty"
        movesHoverText.append(Component.literal("Move 1: ").withStyle(ChatFormatting.RED).append(Component.literal(moveOne).withStyle(ChatFormatting.WHITE)))
        movesHoverText.append(Component.literal("\n"))
        movesHoverText.append(Component.literal("Move 2: ").withStyle(ChatFormatting.YELLOW).append(Component.literal(moveTwo).withStyle(ChatFormatting.WHITE)))
        movesHoverText.append(Component.literal("\n"))
        movesHoverText.append(Component.literal("Move 3: ").withStyle(ChatFormatting.AQUA).append(Component.literal(moveThree).withStyle(ChatFormatting.WHITE)))
        movesHoverText.append(Component.literal("\n"))
        movesHoverText.append(Component.literal("Move 4: ").withStyle(ChatFormatting.GREEN).append(Component.literal(moveFour).withStyle(ChatFormatting.WHITE)))
        movesText.setStyle(movesText.style.withHoverEvent(
            HoverEvent(HoverEvent.Action.SHOW_TEXT,
                movesHoverText
            ))
        )
        toSend.append(movesText)

        return toSend
    }

    fun pokemonToItem(pokemon: Pokemon): ItemStack {
        val moveOne = if (pokemon.moveSet.getMoves().size >= 1) pokemon.moveSet[0]!!.displayName.string else "Empty"
        val moveTwo = if (pokemon.moveSet.getMoves().size >= 2) pokemon.moveSet[1]!!.displayName.string else "Empty"
        val moveThree = if (pokemon.moveSet.getMoves().size >= 3) pokemon.moveSet[2]!!.displayName.string else "Empty"
        val moveFour = if (pokemon.moveSet.getMoves().size >= 4) pokemon.moveSet[3]!!.displayName.string else "Empty"



        val itemstack: ItemStack = ItemBuilder(PokemonItem.from(pokemon,1))
            .hideAdditional()
            .addLore(arrayOf<Component>(Component.literal(pokemon.caughtBall.item().defaultInstance.displayName.string).setStyle(Style.EMPTY.withItalic(true).withColor(ChatFormatting.DARK_GRAY)),
                Component.literal("Level: ").withStyle(ChatFormatting.AQUA).append(Component.literal(pokemon.level.toString()).withStyle(ChatFormatting.WHITE)),

                Component.literal("Nickname: ").withStyle(ChatFormatting.DARK_GREEN).append(Component.literal(
                    pokemon.nickname?.string ?: "No nickname"
                ).withStyle(ChatFormatting.WHITE)),

                Component.literal("Nature: ").withStyle(ChatFormatting.YELLOW).append(lang(pokemon.nature.displayName.replace("cobblemon.", "")).withStyle(ChatFormatting.WHITE)),
                Component.literal("Ability: ").withStyle(ChatFormatting.GOLD).append(lang(pokemon.ability.displayName.replace("cobblemon.", "")).withStyle(ChatFormatting.WHITE)),
                Component.literal("IVs: ").withStyle(ChatFormatting.LIGHT_PURPLE),
                Component.literal("  HP: ").withStyle(ChatFormatting.RED).append(Component.literal(pokemon.ivs.getOrDefault(Stats.HP).toString()).withStyle(ChatFormatting.WHITE))
                    .append(Component.literal("  Atk: ").withStyle(ChatFormatting.BLUE).append(Component.literal(pokemon.ivs.getOrDefault(Stats.ATTACK).toString()).withStyle(ChatFormatting.WHITE)))
                    .append(Component.literal("  Def: ").withStyle(ChatFormatting.GRAY).append(Component.literal(pokemon.ivs.getOrDefault(Stats.DEFENCE).toString()).withStyle(ChatFormatting.WHITE))),
                Component.literal("  SpAtk: ").withStyle(ChatFormatting.AQUA).append(Component.literal(pokemon.ivs.getOrDefault(Stats.SPECIAL_ATTACK).toString()).withStyle(ChatFormatting.WHITE))
                    .append(Component.literal("  SpDef: ").withStyle(ChatFormatting.YELLOW).append(Component.literal(pokemon.ivs.getOrDefault(Stats.SPECIAL_DEFENCE).toString()).withStyle(ChatFormatting.WHITE)))
                    .append(Component.literal("  Spd: ").withStyle(ChatFormatting.GREEN).append(Component.literal(pokemon.ivs.getOrDefault(Stats.SPEED).toString()).withStyle(ChatFormatting.WHITE))),

                Component.literal("EVs: ").withStyle(ChatFormatting.DARK_AQUA),
                Component.literal("  HP: ").withStyle(ChatFormatting.RED).append(Component.literal(pokemon.evs.getOrDefault(Stats.HP).toString()).withStyle(ChatFormatting.WHITE))
                    .append(Component.literal("  Atk: ").withStyle(ChatFormatting.BLUE).append(Component.literal(pokemon.evs.getOrDefault(Stats.ATTACK).toString()).withStyle(ChatFormatting.WHITE)))
                    .append(Component.literal("  Def: ").withStyle(ChatFormatting.GRAY).append(Component.literal(pokemon.evs.getOrDefault(Stats.DEFENCE).toString()).withStyle(ChatFormatting.WHITE))),
                Component.literal("  SpAtk: ").withStyle(ChatFormatting.AQUA).append(Component.literal(pokemon.evs.getOrDefault(Stats.SPECIAL_ATTACK).toString()).withStyle(ChatFormatting.WHITE))
                    .append(Component.literal("  SpDef: ").withStyle(ChatFormatting.YELLOW).append(Component.literal(pokemon.evs.getOrDefault(Stats.SPECIAL_DEFENCE).toString()).withStyle(ChatFormatting.WHITE)))
                    .append(Component.literal("  Spd: ").withStyle(ChatFormatting.GREEN).append(Component.literal(pokemon.evs.getOrDefault(Stats.SPEED).toString()).withStyle(ChatFormatting.WHITE))),
                Component.literal("Moves: ").withStyle(ChatFormatting.DARK_GREEN),
                Component.literal(" ").append(Component.literal(moveOne).withStyle(ChatFormatting.WHITE)),
                Component.literal(" ").append(Component.literal(moveTwo).withStyle(ChatFormatting.WHITE)),
                Component.literal(" ").append(Component.literal(moveThree).withStyle(ChatFormatting.WHITE)),
                Component.literal(" ").append(Component.literal(moveFour).withStyle(ChatFormatting.WHITE)),
                Component.literal("Form: ").withStyle(ChatFormatting.GOLD).append(pokemon.form.name)
            ))
            .setCustomName(
                if (pokemon.shiny) pokemon.species.translatedName.copy().withStyle(ChatFormatting.GRAY).append(Component.literal(" ★").withStyle(ChatFormatting.GOLD)) else pokemon.species.translatedName.copy().withStyle(ChatFormatting.GRAY)
            )
            .build()
        return itemstack
    }
}